// gifted-charts-core/dist/BubbleChart/index.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
import { useState } from 'react';
import { computeMaxAndMinXForBubble, computeMaxAndMinYForBubble, getAxesAndRulesProps, getExtendedContainerHeightWithPadding, getIntegerizedValue, getMaxValue, getNoOfSections, indexOfFirstNonZeroDigit, withinMinMaxRange } from '../utils';
import { AxesAndRulesDefaults, BubbleDefaults, chartTypes, defaultBubbleColors, LineDefaults } from '../utils/constants';
import { Dimensions } from 'react-native';
var screenWidth = Dimensions.get('window').width;
export var useBubbleChart = function (props) {
    var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e;
    var _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54, _55, _56, _57, _58, _59, _60, _61, _62, _63, _64, _65, _66, _67, _68, _69, _70, _71, _72, _73, _74, _75, _76, _77, _78, _79, _80, _81, _82, _83, _84, _85, _86, _87, _88, _89, _90;
    var formatXLabel = props.formatXLabel, dataSet = props.dataSet, _91 = props.disableForeignObject, disableForeignObject = _91 === void 0 ? BubbleDefaults.disableForeignObject : _91;
    var _92 = props.data, data = _92 === void 0 ? [] : _92;
    if (dataSet) {
        var emptyDataSet = [];
        data = dataSet.reduce(function (acc, item) { return __spreadArray(__spreadArray([], __read(acc), false), __read(item.data), false); }, emptyDataSet);
    }
    try {
        // For the sake of uniformity across other charts, allow value as an alternate to y
        for (var data_1 = __values(data), data_1_1 = data_1.next(); !data_1_1.done; data_1_1 = data_1.next()) {
            var item = data_1_1.value;
            if (item.y === undefined && item.value !== undefined)
                item.y = item.value;
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (data_1_1 && !data_1_1.done && (_a = data_1.return)) _a.call(data_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    var yNoOfSections = getNoOfSections(props.yNoOfSections, props.maxY, props.yStepValue);
    var containerHeight = (_f = props.height) !== null && _f !== void 0 ? _f : (((_g = props.yStepHeight) !== null && _g !== void 0 ? _g : 0) * yNoOfSections ||
        AxesAndRulesDefaults.containerHeight);
    var yRange = Math.max.apply(Math, __spreadArray([], __read(data.map(function (i) { return Math.max(i.y, 0); })), false)) - // find the largest +ve number
     Math.min.apply(// find the largest +ve number
    Math, __spreadArray([], __read(data.map(function (i) { return Math.max(i.y, 0); })), false)); // find the smallest +ve number
    var showFractionalYAxis = (_h = props.showFractionalYAxis) !== null && _h !== void 0 ? _h : (isFinite(yRange) && yRange <= 1);
    var roundToDigits = (_j = props.yRoundToDigits) !== null && _j !== void 0 ? _j : (showFractionalYAxis ? indexOfFirstNonZeroDigit(yRange) + 1 : 0);
    var minRadius = (_k = props.minRadius) !== null && _k !== void 0 ? _k : BubbleDefaults.minRadius;
    var maxRadius = (_l = props.maxRadius) !== null && _l !== void 0 ? _l : Math.min(containerHeight, (_m = props.width) !== null && _m !== void 0 ? _m : Infinity) / 5;
    var xRange = Math.max.apply(Math, __spreadArray([], __read(data.map(function (i) { var _a; return Math.max((_a = i.x) !== null && _a !== void 0 ? _a : 0, 0); })), false)) - // find the largest +ve number
     Math.min.apply(// find the largest +ve number
    Math, __spreadArray([], __read(data.map(function (i) { var _a; return Math.max((_a = i.x) !== null && _a !== void 0 ? _a : 0, 0); })), false)) || // find the smallest +ve number
        data.length;
    var xNoOfSections = getNoOfSections(props.xNoOfSections, props.maxX, props.xStepValue, true);
    var initialSpacing = (_o = props.initialSpacing) !== null && _o !== void 0 ? _o : BubbleDefaults.initialSpacing;
    var spacing = (_p = props.spacing) !== null && _p !== void 0 ? _p : (props.width ? props.width / xNoOfSections : LineDefaults.spacing);
    var showFractionalXAxis = (_q = props.showFractionalXAxis) !== null && _q !== void 0 ? _q : (isFinite(xRange) && xRange - 1 <= xNoOfSections);
    var xRoundToDigits = (_r = props.xRoundToDigits) !== null && _r !== void 0 ? _r : (showFractionalXAxis ? indexOfFirstNonZeroDigit(xRange) + 1 : 0);
    var endSpacing = (_s = props.endSpacing) !== null && _s !== void 0 ? _s : BubbleDefaults.endSpacing;
    var totalWidth = initialSpacing + spacing * xNoOfSections + endSpacing;
    var _93 = computeMaxAndMinYForBubble(data, false, // extrapolateMissingValues
    roundToDigits, showFractionalYAxis, data), yMaxItem = _93.maxItem, yMinItem = _93.minItem;
    var bubblesRadius = (_t = props.bubblesRadius) !== null && _t !== void 0 ? _t : BubbleDefaults.bubblesRadius;
    var maxY = // should change only once and that's when autoRoundLabelsY is true
     getMaxValue(props.maxY, props.yStepValue, yNoOfSections, yMaxItem) || 10;
    var mostNegativeY = (_u = props.mostNegativeY) !== null && _u !== void 0 ? _u : yMinItem;
    var onlyPositive = (_v = props.onlyPositive) !== null && _v !== void 0 ? _v : mostNegativeY >= 0;
    var horizSections = [{ value: '0' }];
    var yStepHeight = (_w = props.yStepHeight) !== null && _w !== void 0 ? _w : containerHeight / yNoOfSections;
    var yStepValue = (_x = props.yStepValue) !== null && _x !== void 0 ? _x : maxY / yNoOfSections;
    var yScale = yStepHeight / yStepValue;
    /*******************************************************************************************
     * *************      RE - CALCULATE maxY, yStepValue & yStepHeight               *********
     ******************************************************************************************/
    var topMostReachingBubbleY = 0;
    try {
        for (var data_2 = __values(data), data_2_1 = data_2.next(); !data_2_1.done; data_2_1 = data_2.next()) {
            var item = data_2_1.value;
            var radius = withinMinMaxRange((_y = item.r) !== null && _y !== void 0 ? _y : bubblesRadius, maxRadius, minRadius);
            var maxY_1 = (((_z = item.y) !== null && _z !== void 0 ? _z : 0) + radius / yScale) * 1.01; // 1.01 is a factor to get a little higher value of maxY. This avoids cut-off from top
            topMostReachingBubbleY = Math.max(maxY_1, topMostReachingBubbleY);
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (data_2_1 && !data_2_1.done && (_b = data_2.return)) _b.call(data_2);
        }
        finally { if (e_2) throw e_2.error; }
    }
    maxY = topMostReachingBubbleY;
    yStepHeight = (_0 = props.yStepHeight) !== null && _0 !== void 0 ? _0 : containerHeight / yNoOfSections;
    yStepValue = (_1 = props.yStepValue) !== null && _1 !== void 0 ? _1 : maxY / yNoOfSections;
    /*******************************************************************************************
     *******************************************************************************************/
    var autoRoundLabelsY = (_3 = (_2 = props.autoRoundLabelsY) !== null && _2 !== void 0 ? _2 : props.autoRoundLabels) !== null && _3 !== void 0 ? _3 : BubbleDefaults.autoRoundLabelsY;
    if (autoRoundLabelsY) {
        yStepValue = getIntegerizedValue(yStepValue);
        maxY = yStepValue * yNoOfSections;
    }
    var yNegativeStepValue = (_4 = props.yNegativeStepValue) !== null && _4 !== void 0 ? _4 : yStepValue;
    var xAxisThickness = (_5 = props.xAxisThickness) !== null && _5 !== void 0 ? _5 : AxesAndRulesDefaults.xAxisThickness;
    var _94 = computeMaxAndMinXForBubble(data, false, // extrapolateMissingValues
    xRoundToDigits, showFractionalXAxis, data), xMaxItem = _94.maxItem, xMinItem = _94.minItem;
    var maxX = // should change only once and that's when autoRoundLabelsX is true
     getMaxValue(props.maxX, props.xStepValue, xNoOfSections, xMaxItem) || 10;
    var minX = xMinItem;
    var leftMostReachingBubblesX = Infinity;
    var rightMostReachingBubblesX = 0;
    try {
        for (var _95 = __values(data.entries()), _96 = _95.next(); !_96.done; _96 = _95.next()) {
            var _97 = __read(_96.value, 2), index = _97[0], item = _97[1];
            var radius = withinMinMaxRange((_6 = item.r) !== null && _6 !== void 0 ? _6 : bubblesRadius, maxRadius, minRadius);
            var rescaledRadius = ((_7 = item.x) !== null && _7 !== void 0 ? _7 : index + 1) / (radius === 0 ? 1 : radius);
            var leftX = ((_8 = item.x) !== null && _8 !== void 0 ? _8 : index + 1) - rescaledRadius;
            var rightX = ((_9 = item.x) !== null && _9 !== void 0 ? _9 : index + 1) + rescaledRadius;
            leftMostReachingBubblesX = Math.min(leftX, leftMostReachingBubblesX);
            rightMostReachingBubblesX = Math.max(rightX, rightMostReachingBubblesX);
        }
    }
    catch (e_3_1) { e_3 = { error: e_3_1 }; }
    finally {
        try {
            if (_96 && !_96.done && (_c = _95.return)) _c.call(_95);
        }
        finally { if (e_3) throw e_3.error; }
    }
    var xStepValue = (_10 = props.xStepValue) !== null && _10 !== void 0 ? _10 : (rightMostReachingBubblesX - leftMostReachingBubblesX) / xNoOfSections;
    var noOfSectionsBelowXAxis = (_11 = props.noOfSectionsBelowXAxis) !== null && _11 !== void 0 ? _11 : Math.round(Math.ceil(-mostNegativeY / yNegativeStepValue));
    var labelsExtraHeight = (_12 = props.labelsExtraHeight) !== null && _12 !== void 0 ? _12 : 0;
    var yAxisLabelWidth = (_13 = props.yAxisLabelWidth) !== null && _13 !== void 0 ? _13 : (props.hideYAxisText
        ? AxesAndRulesDefaults.yAxisEmptyLabelWidth
        : AxesAndRulesDefaults.yAxisLabelWidth);
    var horizontal = false;
    var yAxisAtTop = false;
    var disableScroll = (_14 = props.disableScroll) !== null && _14 !== void 0 ? _14 : false;
    var showScrollIndicator = (_15 = props.showScrollIndicator) !== null && _15 !== void 0 ? _15 : false;
    var scrollToEnd = (_16 = props.scrollToEnd) !== null && _16 !== void 0 ? _16 : false;
    var scrollAnimation = (_17 = props.scrollAnimation) !== null && _17 !== void 0 ? _17 : false;
    var scrollEventThrottle = (_18 = props.scrollEventThrottle) !== null && _18 !== void 0 ? _18 : 16;
    var focusEnabled = (_19 = props.focusEnabled) !== null && _19 !== void 0 ? _19 : LineDefaults.focusEnabled;
    var showBubbleOnFocus = (_20 = props.showBubbleOnFocus) !== null && _20 !== void 0 ? _20 : BubbleDefaults.showBubbleOnFocus;
    var _98 = __read(useState((_21 = props.focusedBubbleIndex) !== null && _21 !== void 0 ? _21 : -1), 2), selectedIndex = _98[0], setSelectedIndex = _98[1];
    var _99 = __read(useState(''), 2), points = _99[0], setPoints = _99[1];
    var _100 = __read(useState(-1), 2), pointerIndex = _100[0], setPointerIndex = _100[1];
    var _101 = __read(useState(0), 2), pointerX = _101[0], setPointerX = _101[1];
    var _102 = __read(useState(0), 2), pointerY = _102[0], setPointerY = _102[1];
    var isAnimated = (_22 = props.isAnimated) !== null && _22 !== void 0 ? _22 : false;
    var horizontalRulesStyle = props.horizontalRulesStyle;
    var yAxisOffset = (_23 = props.yAxisOffset) !== null && _23 !== void 0 ? _23 : 0;
    var showXAxisIndices = (_24 = props.showXAxisIndices) !== null && _24 !== void 0 ? _24 : AxesAndRulesDefaults.showXAxisIndices;
    var xAxisIndicesHeight = (_25 = props.xAxisIndicesHeight) !== null && _25 !== void 0 ? _25 : AxesAndRulesDefaults.xAxisIndicesHeight;
    var xAxisIndicesWidth = (_26 = props.xAxisIndicesWidth) !== null && _26 !== void 0 ? _26 : AxesAndRulesDefaults.xAxisIndicesWidth;
    var xAxisIndicesColor = (_27 = props.xAxisIndicesColor) !== null && _27 !== void 0 ? _27 : AxesAndRulesDefaults.xAxisIndicesColor;
    var getPointerProps = (_28 = props.getPointerProps) !== null && _28 !== void 0 ? _28 : null;
    var bubblesWidth = (_29 = props.bubblesWidth) !== null && _29 !== void 0 ? _29 : BubbleDefaults.bubblesWidth;
    var extraWidthDueToBubble = props.hideBubbles
        ? 0
        : Math.min(10, bubblesRadius !== null && bubblesRadius !== void 0 ? bubblesRadius : bubblesWidth);
    var xAxisLabelsAtBottom = (_30 = props.xAxisLabelsAtBottom) !== null && _30 !== void 0 ? _30 : false;
    var allowFontScaling = (_31 = props.allowFontScaling) !== null && _31 !== void 0 ? _31 : AxesAndRulesDefaults.allowFontScaling;
    var animationDuration = (_32 = props.animationDuration) !== null && _32 !== void 0 ? _32 : LineDefaults.animationDuration;
    var secondaryMaxValue = maxY;
    var secondaryMinItem = maxY * -1;
    var showSecondaryFractionalValues = false;
    var secondaryRoundToDigits = 1;
    // const autoRoundLabelsX =
    //   props.autoRoundLabelsX ??
    //   props.autoRoundLabels ??
    //   BubbleDefaults.autoRoundLabelsX
    var axesAndRulesProps = getAxesAndRulesProps(props, containerHeight, yStepHeight, yStepValue, yNoOfSections, roundToDigits, yNegativeStepValue, secondaryMaxValue, secondaryMinItem, showSecondaryFractionalValues, secondaryRoundToDigits);
    var fourthQuadrantHeight = noOfSectionsBelowXAxis * yStepHeight;
    var extendedContainerHeight = getExtendedContainerHeightWithPadding(containerHeight, 0);
    var containerHeightIncludingBelowXAxis = extendedContainerHeight + fourthQuadrantHeight;
    var mostNegativeValueOnYAxis = yNegativeStepValue * noOfSectionsBelowXAxis;
    var containsNegativeValue = ((_33 = props.mostNegativeY) !== null && _33 !== void 0 ? _33 : 0) < 0 || (data === null || data === void 0 ? void 0 : data.some(function (item) { var _a; return ((_a = item.y) !== null && _a !== void 0 ? _a : 0) < 0; }));
    /*******************************************************************************************
     * *************               RE - CALCULATE xScale & xStepValue                  *********
     ******************************************************************************************/
    var xScale = spacing / xStepValue;
    leftMostReachingBubblesX = Infinity;
    rightMostReachingBubblesX = 0;
    try {
        for (var _103 = __values(data.entries()), _104 = _103.next(); !_104.done; _104 = _103.next()) {
            var _105 = __read(_104.value, 2), index = _105[0], item = _105[1];
            var radius = withinMinMaxRange((_34 = item.r) !== null && _34 !== void 0 ? _34 : bubblesRadius, maxRadius, minRadius);
            var leftX = ((_35 = item.x) !== null && _35 !== void 0 ? _35 : index + 1) - radius / xScale;
            var rightX = ((_36 = item.x) !== null && _36 !== void 0 ? _36 : index + 1) + radius / xScale;
            leftMostReachingBubblesX = Math.min(leftX, leftMostReachingBubblesX);
            rightMostReachingBubblesX = Math.max(rightX, rightMostReachingBubblesX);
        }
    }
    catch (e_4_1) { e_4 = { error: e_4_1 }; }
    finally {
        try {
            if (_104 && !_104.done && (_d = _103.return)) _d.call(_103);
        }
        finally { if (e_4) throw e_4.error; }
    }
    xStepValue =
        (_37 = props.xStepValue) !== null && _37 !== void 0 ? _37 : (rightMostReachingBubblesX - leftMostReachingBubblesX) / xNoOfSections;
    xScale = spacing / xStepValue;
    /*******************************************************************************************
     *******************************************************************************************/
    /*******************************************************************************************
     * *************            3rd iteration for xScale & xStepValue             *********
     ******************************************************************************************/
    leftMostReachingBubblesX = Infinity;
    rightMostReachingBubblesX = 0;
    try {
        for (var _106 = __values(data.entries()), _107 = _106.next(); !_107.done; _107 = _106.next()) {
            var _108 = __read(_107.value, 2), index = _108[0], item = _108[1];
            var radius = withinMinMaxRange((_38 = item.r) !== null && _38 !== void 0 ? _38 : bubblesRadius, maxRadius, minRadius);
            var leftX = ((_39 = item.x) !== null && _39 !== void 0 ? _39 : index + 1) - radius / xScale;
            var rightX = ((_40 = item.x) !== null && _40 !== void 0 ? _40 : index + 1) + radius / xScale;
            leftMostReachingBubblesX = Math.min(leftX, leftMostReachingBubblesX);
            rightMostReachingBubblesX = Math.max(rightX, rightMostReachingBubblesX);
        }
    }
    catch (e_5_1) { e_5 = { error: e_5_1 }; }
    finally {
        try {
            if (_107 && !_107.done && (_e = _106.return)) _e.call(_106);
        }
        finally { if (e_5) throw e_5.error; }
    }
    // this statement is to polish the leftMostReachingBubblesX to avoid being cut off by x axis
    leftMostReachingBubblesX -=
        Number((rightMostReachingBubblesX - leftMostReachingBubblesX) / xNoOfSections //.toFixed(roundToDigits)
        ) / 5;
    xStepValue =
        (_41 = props.xStepValue) !== null && _41 !== void 0 ? _41 : Number((rightMostReachingBubblesX - leftMostReachingBubblesX) / xNoOfSections //.toFixed(roundToDigits)
        );
    xStepValue += xStepValue / 20; // this statement is to polish thexStepValue to avoid the leftmost bubble being cut off by x axis
    // if (autoRoundLabelsX) {
    //   xStepValue = getIntegerizedValue(xStepValue)
    //   maxX = xStepValue * xNoOfSections
    // }
    xScale = spacing / xStepValue;
    /*******************************************************************************************
     *******************************************************************************************/
    var initialX = (_42 = props.xAxisOffset) !== null && _42 !== void 0 ? _42 : Math.max((_43 = props.minX) !== null && _43 !== void 0 ? _43 : -Infinity, leftMostReachingBubblesX);
    var xAxisLabelTexts = (_44 = props.xAxisLabelTexts) !== null && _44 !== void 0 ? _44 : Array.from({ length: xNoOfSections + 1 }, function (_, i) {
        var labelText = (initialX + xStepValue * i)
            .toFixed(xRoundToDigits)
            .toString();
        if (formatXLabel) {
            return formatXLabel(labelText);
        }
        return labelText;
    });
    var getX = function (index, spreadIndex) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        var item = data === null || data === void 0 ? void 0 : data[spreadIndex !== null && spreadIndex !== void 0 ? spreadIndex : index];
        var val;
        if ((item === null || item === void 0 ? void 0 : item.x) !== undefined) {
            val = (((_a = item === null || item === void 0 ? void 0 : item.x) !== null && _a !== void 0 ? _a : 0) - initialX) * xScale + initialSpacing;
        }
        else {
            var radius = withinMinMaxRange((_b = item === null || item === void 0 ? void 0 : item.r) !== null && _b !== void 0 ? _b : bubblesRadius, maxRadius, minRadius);
            val =
                Math.min(totalWidth - ((_c = item === null || item === void 0 ? void 0 : item.r) !== null && _c !== void 0 ? _c : BubbleDefaults.bubblesRadius), (index * totalWidth) /
                    ((_h = (_g = ((_f = (_e = (_d = props.dataSet) === null || _d === void 0 ? void 0 : _d[0]) === null || _e === void 0 ? void 0 : _e.data) !== null && _f !== void 0 ? _f : props.data)) === null || _g === void 0 ? void 0 : _g.length) !== null && _h !== void 0 ? _h : 1)) -
                    initialX +
                    radius;
        }
        return val;
    };
    var getY = function (value) {
        if (containsNegativeValue &&
            value < 0 &&
            yStepValue !== yNegativeStepValue) {
            return (extendedContainerHeight +
                (value * fourthQuadrantHeight) / mostNegativeValueOnYAxis);
        }
        return extendedContainerHeight - (value * containerHeight) / maxY;
    };
    var showTextOnFocus = (_45 = props.showTextOnFocus) !== null && _45 !== void 0 ? _45 : LineDefaults.showTextOnFocus;
    var showBubbleLabelOnFocus = (_46 = props.showBubbleLabelOnFocus) !== null && _46 !== void 0 ? _46 : BubbleDefaults.showBubbleLabelOnFocus;
    var focusTogether = (_47 = props.focusTogether) !== null && _47 !== void 0 ? _47 : true;
    var _109 = __read(useState(-1), 2), selectedLineNumber = _109[0], setSelectedLineNumber = _109[1];
    var lastLineNumber = 1;
    var unFocusOnPressOut = (_48 = props.unFocusOnPressOut) !== null && _48 !== void 0 ? _48 : LineDefaults.unFocusOnPressOut;
    var delayBeforeUnFocus = (_49 = props.delayBeforeUnFocus) !== null && _49 !== void 0 ? _49 : LineDefaults.delayBeforeUnFocus;
    var handleFocus = function (index) {
        setSelectedIndex(index);
    };
    var handleUnFocus = function () {
        if (unFocusOnPressOut) {
            setTimeout(function () { return setSelectedIndex(-1); }, delayBeforeUnFocus);
        }
    };
    var scatterChart = (_50 = props.scatterChart) !== null && _50 !== void 0 ? _50 : BubbleDefaults.scatterChart;
    var hideBubbles = (_51 = props.hideBubbles) !== null && _51 !== void 0 ? _51 : false;
    var bubblesShape = (_52 = props.bubblesShape) !== null && _52 !== void 0 ? _52 : BubbleDefaults.bubblesShape;
    var bubblesHeight = (_53 = props.bubblesHeight) !== null && _53 !== void 0 ? _53 : BubbleDefaults.bubblesHeight;
    var bubblesColor = scatterChart
        ? ((_54 = props.bubblesColor) !== null && _54 !== void 0 ? _54 : BubbleDefaults.bubblesColor)
        : props.bubblesColor;
    var startIndex = 0;
    var endIndex = (_59 = (_58 = ((_57 = (_56 = (_55 = props.dataSet) === null || _55 === void 0 ? void 0 : _55[0]) === null || _56 === void 0 ? void 0 : _56.data) !== null && _57 !== void 0 ? _57 : props.data)) === null || _58 === void 0 ? void 0 : _58.length) !== null && _59 !== void 0 ? _59 : 0;
    var labelFontSize = (_60 = props.labelFontSize) !== null && _60 !== void 0 ? _60 : BubbleDefaults.labelFontSize;
    var labelMaxLength = Math.max(3, (_61 = props.labelMaxLength) !== null && _61 !== void 0 ? _61 : BubbleDefaults.labelMaxLength);
    var showValuesAsBubbleLabels = (_62 = props.showValuesAsBubbleLabels) !== null && _62 !== void 0 ? _62 : BubbleDefaults.showValuesAsBubbleLabels;
    var xAxisLabelsVerticalShift = (_63 = props.xAxisLabelsVerticalShift) !== null && _63 !== void 0 ? _63 : (xAxisLabelsAtBottom
        ? fourthQuadrantHeight
        : AxesAndRulesDefaults.xAxisLabelsVerticalShift);
    var xAxisTextNumberOfLines = (_64 = props.xAxisTextNumberOfLines) !== null && _64 !== void 0 ? _64 : AxesAndRulesDefaults.xAxisTextNumberOfLines;
    var rotateLabel = (_65 = props.rotateLabel) !== null && _65 !== void 0 ? _65 : false;
    var borderWidth = (_66 = props.borderWidth) !== null && _66 !== void 0 ? _66 : BubbleDefaults.borderWidth;
    var borderColor = scatterChart
        ? ((_67 = props.borderColor) !== null && _67 !== void 0 ? _67 : BubbleDefaults.borderColor)
        : props.borderColor;
    var opacity = (_68 = props.opacity) !== null && _68 !== void 0 ? _68 : BubbleDefaults.opacity;
    var borderOpacity = (_69 = props.borderOpacity) !== null && _69 !== void 0 ? _69 : BubbleDefaults.borderOpacity;
    var showRegressionLine = (_70 = props.showRegressionLine) !== null && _70 !== void 0 ? _70 : false;
    var weightedRegression = function (data) {
        var e_6, _a;
        if (!data)
            return { slope: 0, intercept: 0 };
        var sumW = 0;
        var sumWX = 0;
        var sumWY = 0;
        var sumWXX = 0;
        var sumWXY = 0;
        var i = -1;
        try {
            for (var data_3 = __values(data), data_3_1 = data_3.next(); !data_3_1.done; data_3_1 = data_3.next()) {
                var _b = data_3_1.value, y = _b.y, _c = _b.r, r = _c === void 0 ? 1 : _c;
                ++i;
                var w = r * r;
                if (w <= 0)
                    continue; // ignore zero/negative weights
                var x = getX(i);
                sumW += w;
                sumWX += w * x;
                sumWY += w * y;
                sumWXX += w * x * x;
                sumWXY += w * x * y;
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (data_3_1 && !data_3_1.done && (_a = data_3.return)) _a.call(data_3);
            }
            finally { if (e_6) throw e_6.error; }
        }
        var denominator = sumW * sumWXX - sumWX * sumWX;
        if (denominator === 0) {
            throw new Error('Regression not defined (degenerate data)');
        }
        var slope = (sumW * sumWXY - sumWX * sumWY) / denominator;
        var intercept = (sumWY - slope * sumWX) / sumW;
        return { slope: slope, intercept: intercept };
    };
    var getAdjustedScaledX = function (x) {
        return (x - initialX) * xScale + initialSpacing;
    };
    var regressionLineConfig = {
        thickness: (_72 = (_71 = props.regressionLineConfig) === null || _71 === void 0 ? void 0 : _71.thickness) !== null && _72 !== void 0 ? _72 : BubbleDefaults.regressionLineConfig.thickness,
        color: (_74 = (_73 = props.regressionLineConfig) === null || _73 === void 0 ? void 0 : _73.color) !== null && _74 !== void 0 ? _74 : BubbleDefaults.regressionLineConfig.color,
        opacity: (_76 = (_75 = props.regressionLineConfig) === null || _75 === void 0 ? void 0 : _75.opacity) !== null && _76 !== void 0 ? _76 : BubbleDefaults.regressionLineConfig.opacity,
        strokeDashArray: (_77 = props.regressionLineConfig) === null || _77 === void 0 ? void 0 : _77.strokeDashArray,
        isAnimated: (_79 = (_78 = props.regressionLineConfig) === null || _78 === void 0 ? void 0 : _78.isAnimated) !== null && _79 !== void 0 ? _79 : isAnimated,
        animationDuration: (_81 = (_80 = props.regressionLineConfig) === null || _80 === void 0 ? void 0 : _80.animationDuration) !== null && _81 !== void 0 ? _81 : animationDuration,
        x1: (_82 = props.regressionLineConfig) === null || _82 === void 0 ? void 0 : _82.x1,
        x2: (_83 = props.regressionLineConfig) === null || _83 === void 0 ? void 0 : _83.x2,
        y1: (_84 = props.regressionLineConfig) === null || _84 === void 0 ? void 0 : _84.y1,
        y2: (_85 = props.regressionLineConfig) === null || _85 === void 0 ? void 0 : _85.y2
    };
    var regressionLineCoordinates = [];
    var regressionLineConfigs = [];
    var regressionLineX1 = 0, regressionLineY1 = 0, regressionLineX2 = 0, regressionLineY2 = 0;
    if (dataSet) {
        dataSet === null || dataSet === void 0 ? void 0 : dataSet.forEach(function (item) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
            /************           Add default x values to each data item in case of DataSet           ************/
            item.data.forEach(function (dataItem, index) {
                dataItem.indexUsedInDevForDataSet = index;
                if (item.seriesBubblesColor) {
                    dataItem.bubbleColor = item.seriesBubblesColor;
                }
                else if (!bubblesColor && !dataItem.bubbleColor) {
                    dataItem.bubbleColor =
                        defaultBubbleColors[index % defaultBubbleColors.length];
                }
                if (item.seriesBubblesBorderColor) {
                    dataItem.borderColor = item.seriesBubblesBorderColor;
                }
                else if (!borderColor && !dataItem.borderColor) {
                    dataItem.borderColor =
                        defaultBubbleColors[index % defaultBubbleColors.length];
                }
            });
            /*******************************************************************************************************/
            if (item.showRegressionLine) {
                var currentRegressionLineConfig = item.regressionLineConfig;
                var currentRegressionLineConfigModified = {
                    thickness: (_a = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.thickness) !== null && _a !== void 0 ? _a : regressionLineConfig.thickness,
                    color: (_b = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.color) !== null && _b !== void 0 ? _b : regressionLineConfig.color,
                    opacity: (_c = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.opacity) !== null && _c !== void 0 ? _c : regressionLineConfig.opacity,
                    strokeDashArray: (_d = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.strokeDashArray) !== null && _d !== void 0 ? _d : regressionLineConfig.strokeDashArray,
                    isAnimated: (_e = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.isAnimated) !== null && _e !== void 0 ? _e : regressionLineConfig.isAnimated,
                    animationDuration: (_f = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.animationDuration) !== null && _f !== void 0 ? _f : regressionLineConfig.animationDuration,
                    x1: (_g = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.x1) !== null && _g !== void 0 ? _g : regressionLineConfig.x1,
                    x2: (_h = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.x2) !== null && _h !== void 0 ? _h : regressionLineConfig.x2,
                    y1: (_j = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.y1) !== null && _j !== void 0 ? _j : regressionLineConfig.y1,
                    y2: (_k = currentRegressionLineConfig === null || currentRegressionLineConfig === void 0 ? void 0 : currentRegressionLineConfig.y2) !== null && _k !== void 0 ? _k : regressionLineConfig.y2
                };
                regressionLineConfigs.push(currentRegressionLineConfigModified);
                var regressionLineX1_1 = 0, regressionLineY1_1 = 0, regressionLineX2_1 = 0, regressionLineY2_1 = 0;
                var slope = 0, intercept = 0;
                var slopeIntercept = weightedRegression(item.data);
                slope = slopeIntercept.slope;
                intercept = slopeIntercept.intercept;
                // Calculate Y values in data space first
                var y1_data = slope * minX + intercept;
                var y2_data = slope * maxX + intercept;
                // Convert X coordinates to screen space
                regressionLineX1_1 = getAdjustedScaledX((_l = currentRegressionLineConfigModified.x1) !== null && _l !== void 0 ? _l : 0);
                regressionLineX2_1 =
                    currentRegressionLineConfigModified.x2 !== undefined
                        ? getAdjustedScaledX(currentRegressionLineConfigModified.x2)
                        : Math.min(screenWidth, (_m = props.width) !== null && _m !== void 0 ? _m : totalWidth);
                // Convert Y coordinates to screen space using getY function
                regressionLineY1_1 =
                    currentRegressionLineConfigModified.y1 !== undefined
                        ? getY(currentRegressionLineConfigModified.y1)
                        : getY(y1_data);
                regressionLineY2_1 =
                    currentRegressionLineConfigModified.y2 !== undefined
                        ? getY(currentRegressionLineConfigModified.y2)
                        : getY(y2_data);
                regressionLineCoordinates.push({
                    regressionLineX1: regressionLineX1_1,
                    regressionLineX2: regressionLineX2_1,
                    regressionLineY1: regressionLineY1_1,
                    regressionLineY2: regressionLineY2_1
                });
            }
        });
    }
    else {
        if (showRegressionLine) {
            var slope = 0, intercept = 0;
            var slopeIntercept = weightedRegression(data);
            slope = slopeIntercept.slope;
            intercept = slopeIntercept.intercept;
            // Calculate Y values in data space first
            var y1_data = slope * minX + intercept;
            var y2_data = slope * maxX + intercept;
            // Convert X coordinates to screen space
            regressionLineX1 = getAdjustedScaledX((_86 = regressionLineConfig.x1) !== null && _86 !== void 0 ? _86 : 0);
            regressionLineX2 =
                regressionLineConfig.x2 !== undefined
                    ? getAdjustedScaledX(regressionLineConfig.x2)
                    : Math.min(screenWidth, (_87 = props.width) !== null && _87 !== void 0 ? _87 : totalWidth);
            // Convert Y coordinates to screen space using getY function
            regressionLineY1 = regressionLineConfig.y1
                ? getY(regressionLineConfig.y1)
                : getY(y1_data);
            regressionLineY2 = regressionLineConfig.y2
                ? getY(regressionLineConfig.y2)
                : getY(y2_data);
        }
    }
    var showGradient = (_88 = props.showGradient) !== null && _88 !== void 0 ? _88 : (scatterChart ? false : BubbleDefaults.showGradient);
    var centerColorForGradient = (_89 = props.centerColorForGradient) !== null && _89 !== void 0 ? _89 : BubbleDefaults.centerColorForGradient;
    var barAndLineChartsWrapperProps = {
        chartType: chartTypes.BUBBLE,
        containerHeight: containerHeight,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: yStepHeight,
        negativeStepHeight: yStepHeight,
        labelsExtraHeight: labelsExtraHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        rtl: false,
        shiftX: 0,
        shiftY: 0,
        yAxisAtTop: yAxisAtTop,
        initialSpacing: initialSpacing,
        data: data,
        stackData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        secondaryData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        barWidth: 0, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        xAxisThickness: xAxisThickness,
        totalWidth: totalWidth,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollToIndex: props.scrollToIndex,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        indicatorColor: props.indicatorColor,
        selectedIndex: [selectedIndex],
        setSelectedIndex: setSelectedIndex,
        spacing: spacing,
        showLine: false,
        lineConfig: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineConfig2: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        maxValue: maxY,
        lineData: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineData2: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineBehindBars: false,
        points: points,
        points2: '', // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        arrowPoints: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        // horizSectionProps-
        width: props.width,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: yNoOfSections,
        sectionColors: props.ySectionColors,
        showFractionalValues: showFractionalYAxis,
        axesAndRulesProps: axesAndRulesProps,
        yAxisLabelTexts: props.yAxisLabelTexts,
        yAxisOffset: yAxisOffset,
        rotateYAxisTexts: 0,
        hideAxesAndRules: props.hideAxesAndRules,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        getPointerProps: getPointerProps,
        pointerIndex: pointerIndex,
        pointerX: pointerX,
        pointerY: pointerY,
        onEndReached: props.onEndReached,
        onStartReached: props.onStartReached,
        endReachedOffset: (_90 = props.endReachedOffset) !== null && _90 !== void 0 ? _90 : LineDefaults.endReachedOffset,
        onMomentumScrollEnd: props.onMomentumScrollEnd,
        extraWidthDueToDataPoint: extraWidthDueToBubble,
        customBackground: props.customBackground,
        onlyPositive: onlyPositive,
        highlightEnabled: LineDefaults.highlightEnabled,
        lowlightOpacity: LineDefaults.lowlightOpacity,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        onScrollEndDrag: props.onScrollEndDrag,
        allowFontScaling: allowFontScaling,
        showVerticalLines: props.showVerticalLines,
        xAxisLabelTexts: xAxisLabelTexts,
        disableForeignObject: disableForeignObject
    };
    return {
        data: data,
        totalWidth: totalWidth,
        animationDuration: animationDuration,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        getY: getY,
        barAndLineChartsWrapperProps: barAndLineChartsWrapperProps,
        getX: getX,
        maxValue: maxY,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        showTextOnFocus: showTextOnFocus,
        focusEnabled: focusEnabled,
        focusTogether: focusTogether,
        selectedLineNumber: selectedLineNumber,
        lastLineNumber: lastLineNumber,
        initialSpacing: initialSpacing,
        spacing: spacing,
        containerHeight: containerHeight,
        handleFocus: handleFocus,
        handleUnFocus: handleUnFocus,
        isAnimated: isAnimated,
        showBubbleOnFocus: showBubbleOnFocus,
        showBubbleLabelOnFocus: showBubbleLabelOnFocus,
        bubblesShape: bubblesShape,
        bubblesWidth: bubblesWidth,
        bubblesHeight: bubblesHeight,
        bubblesColor: bubblesColor,
        bubblesRadius: bubblesRadius,
        minRadius: minRadius,
        maxRadius: maxRadius,
        labelFontSize: labelFontSize,
        labelMaxLength: labelMaxLength,
        labelTextStyle: props.labelTextStyle,
        startIndex: startIndex,
        endIndex: endIndex,
        showValuesAsBubbleLabels: showValuesAsBubbleLabels,
        hideBubbles: hideBubbles,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        labelsExtraHeight: labelsExtraHeight,
        xAxisThickness: xAxisThickness,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        rotateLabel: rotateLabel,
        allowFontScaling: allowFontScaling,
        borderWidth: borderWidth,
        borderColor: borderColor,
        opacity: opacity,
        borderOpacity: borderOpacity,
        xAxisLabelTexts: xAxisLabelTexts,
        showRegressionLine: showRegressionLine,
        regressionLineConfig: regressionLineConfig,
        regressionLineConfigs: regressionLineConfigs,
        regressionLineX1: regressionLineX1,
        regressionLineY1: regressionLineY1,
        regressionLineX2: regressionLineX2,
        regressionLineY2: regressionLineY2,
        regressionLineCoordinates: regressionLineCoordinates,
        scatterChart: scatterChart,
        extraWidthDueToBubble: extraWidthDueToBubble,
        showGradient: showGradient,
        centerColorForGradient: centerColorForGradient,
        disableForeignObject: disableForeignObject
    };
};
