// gifted-charts-core/dist/components/BarAndLineChartsWrapper/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useEffect, useState } from 'react';
import { AxesAndRulesDefaults, BarDefaults, chartTypes } from '../../utils/constants';
import { computeMaxAndMinItems } from '../../utils';
export var useBarAndLineChartsWrapper = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21;
    var chartType = props.chartType, containerHeight = props.containerHeight, noOfSectionsBelowXAxis = props.noOfSectionsBelowXAxis, sectionColors = props.sectionColors, stepHeight = props.stepHeight, negativeStepHeight = props.negativeStepHeight, labelsExtraHeight = props.labelsExtraHeight, yAxisLabelWidth = props.yAxisLabelWidth, horizontal = props.horizontal, rtl = props.rtl, shiftX = props.shiftX, shiftY = props.shiftY, initialSpacing = props.initialSpacing, data = props.data, dataSet = props.dataSet, stackData = props.stackData, secondaryData = props.secondaryData, barWidth = props.barWidth, xAxisThickness = props.xAxisThickness, totalWidth = props.totalWidth, spacing = props.spacing, lineConfig = props.lineConfig, lineConfig2 = props.lineConfig2, maxValue = props.maxValue, lineData = props.lineData, lineData2 = props.lineData2, animatedWidth = props.animatedWidth, lineBehindBars = props.lineBehindBars, points = props.points, points2 = props.points2, arrowPoints = props.arrowPoints, width = props.width, horizSections = props.horizSections, endSpacing = props.endSpacing, horizontalRulesStyle = props.horizontalRulesStyle, noOfSections = props.noOfSections, showFractionalValues = props.showFractionalValues, axesAndRulesProps = props.axesAndRulesProps, yAxisLabelTexts = props.yAxisLabelTexts, yAxisOffset = props.yAxisOffset, rotateYAxisTexts = props.rotateYAxisTexts, pointerConfig = props.pointerConfig, getPointerProps = props.getPointerProps, pointerIndex = props.pointerIndex, pointerX = props.pointerX, pointerY = props.pointerY, scrollEventThrottle = props.scrollEventThrottle, endReachedOffset = props.endReachedOffset, isRTL = props.isRTL, selectedIndex = props.selectedIndex, onlyPositive = props.onlyPositive, highlightEnabled = props.highlightEnabled, lowlightOpacity = props.lowlightOpacity, xAxisLabelsAtBottom = props.xAxisLabelsAtBottom, floatingYAxisLabels = props.floatingYAxisLabels, allowFontScaling = props.allowFontScaling, xAxisLabelTexts = props.xAxisLabelTexts, disableForeignObject = props.disableForeignObject;
    var _22 = (_a = axesAndRulesProps.secondaryYAxisConfig) !== null && _a !== void 0 ? _a : {}, secondaryStepValue = _22.stepValue, secondaryNegativeStepValue = _22.negativeStepValue, secondaryNoOfSectionsBelowXAxis = _22.noOfSectionsBelowXAxis, showSecondaryFractionalValues = _22.showFractionalValues, secondaryRoundToDigits = _22.roundToDigits, secondaryStepHeight = _22.stepHeight, secondaryNegativeStepHeight = _22.negativeStepHeight;
    var primaryYAxisHeightBelowOrigin = noOfSectionsBelowXAxis * negativeStepHeight;
    var secondaryYAxisHeightBelowOrigin = secondaryNoOfSectionsBelowXAxis * secondaryNegativeStepHeight;
    var biggerNegativeYAxisHeight = Math.max(primaryYAxisHeightBelowOrigin, secondaryYAxisHeightBelowOrigin);
    var yAxisAtTop = rtl ? !props.yAxisAtTop : props.yAxisAtTop;
    var hideOrigin = (_b = axesAndRulesProps.hideOrigin) !== null && _b !== void 0 ? _b : AxesAndRulesDefaults.hideOrigin;
    var yAxisSide = (_c = axesAndRulesProps.yAxisSide) !== null && _c !== void 0 ? _c : AxesAndRulesDefaults.yAxisSide;
    var yAxisLabelContainerStyle = axesAndRulesProps.yAxisLabelContainerStyle;
    var yAxisColor = (_d = axesAndRulesProps.yAxisColor) !== null && _d !== void 0 ? _d : AxesAndRulesDefaults.yAxisColor;
    var yAxisExtraHeight = (_e = axesAndRulesProps.yAxisExtraHeight) !== null && _e !== void 0 ? _e : containerHeight / 20;
    var trimYAxisAtTop = (_f = axesAndRulesProps.trimYAxisAtTop) !== null && _f !== void 0 ? _f : AxesAndRulesDefaults.trimYAxisAtTop;
    var overflowTop = axesAndRulesProps.overflowTop;
    var yAxisThickness = (_g = axesAndRulesProps.yAxisThickness) !== null && _g !== void 0 ? _g : AxesAndRulesDefaults.yAxisThickness;
    var xAxisColor = (_h = axesAndRulesProps.xAxisColor) !== null && _h !== void 0 ? _h : AxesAndRulesDefaults.xAxisColor;
    var xAxisLength = axesAndRulesProps.xAxisLength;
    var xAxisType = (_j = axesAndRulesProps.xAxisType) !== null && _j !== void 0 ? _j : AxesAndRulesDefaults.xAxisType;
    var xAxisLabelsVerticalShift = (_k = axesAndRulesProps.xAxisLabelsVerticalShift) !== null && _k !== void 0 ? _k : (chartType === chartTypes.LINE && xAxisLabelsAtBottom
        ? biggerNegativeYAxisHeight
        : AxesAndRulesDefaults.xAxisLabelsVerticalShift);
    var xAxisLabelsHeight = axesAndRulesProps.xAxisLabelsHeight;
    var xAxisTextNumberOfLines = axesAndRulesProps.xAxisTextNumberOfLines;
    var dashWidth = (_l = axesAndRulesProps.dashWidth) !== null && _l !== void 0 ? _l : AxesAndRulesDefaults.dashWidth;
    var dashGap = (_m = axesAndRulesProps.dashGap) !== null && _m !== void 0 ? _m : AxesAndRulesDefaults.dashGap;
    var backgroundColor = (_o = axesAndRulesProps.backgroundColor) !== null && _o !== void 0 ? _o : AxesAndRulesDefaults.backgroundColor;
    var hideRules = (_p = axesAndRulesProps.hideRules) !== null && _p !== void 0 ? _p : AxesAndRulesDefaults.hideRules;
    var rulesLength = axesAndRulesProps.rulesLength;
    var rulesType = (_q = axesAndRulesProps.rulesType) !== null && _q !== void 0 ? _q : AxesAndRulesDefaults.rulesType;
    var rulesThickness = (_r = axesAndRulesProps.rulesThickness) !== null && _r !== void 0 ? _r : AxesAndRulesDefaults.rulesThickness;
    var rulesColor = (_s = axesAndRulesProps.rulesColor) !== null && _s !== void 0 ? _s : AxesAndRulesDefaults.rulesColor;
    var rulesConfigArray = (_t = axesAndRulesProps.rulesConfigArray) !== null && _t !== void 0 ? _t : AxesAndRulesDefaults.rulesConfigArray;
    var showYAxisIndices = (_u = axesAndRulesProps.showYAxisIndices) !== null && _u !== void 0 ? _u : false;
    var yAxisIndicesHeight = (_v = axesAndRulesProps.yAxisIndicesHeight) !== null && _v !== void 0 ? _v : AxesAndRulesDefaults.yAxisIndicesHeight;
    var yAxisIndicesWidth = (_w = axesAndRulesProps.yAxisIndicesWidth) !== null && _w !== void 0 ? _w : AxesAndRulesDefaults.yAxisIndicesWidth;
    var yAxisIndicesColor = (_x = axesAndRulesProps.yAxisIndicesColor) !== null && _x !== void 0 ? _x : AxesAndRulesDefaults.yAxisIndicesColor;
    var hideYAxisText = (_y = axesAndRulesProps.hideYAxisText) !== null && _y !== void 0 ? _y : AxesAndRulesDefaults.hideYAxisText;
    var yAxisTextNumberOfLines = (_z = axesAndRulesProps.yAxisTextNumberOfLines) !== null && _z !== void 0 ? _z : AxesAndRulesDefaults.yAxisTextNumberOfLines;
    var yAxisLabelPrefix = (_0 = axesAndRulesProps.yAxisLabelPrefix) !== null && _0 !== void 0 ? _0 : '';
    var yAxisLabelSuffix = (_1 = axesAndRulesProps.yAxisLabelSuffix) !== null && _1 !== void 0 ? _1 : '';
    var yAxisTextStyle = axesAndRulesProps.yAxisTextStyle;
    var secondaryYAxis = axesAndRulesProps.secondaryYAxis;
    var stepValue = axesAndRulesProps.stepValue;
    var negativeStepValue = axesAndRulesProps.negativeStepValue;
    var roundToDigits = axesAndRulesProps.roundToDigits;
    var referenceLinesConfig = (_2 = axesAndRulesProps.referenceLinesConfig) !== null && _2 !== void 0 ? _2 : {};
    var referenceLinesOverChartContent = (_3 = referenceLinesConfig.referenceLinesOverChartContent) !== null && _3 !== void 0 ? _3 : AxesAndRulesDefaults.referenceLinesOverChartContent;
    var showVerticalLines = (_4 = axesAndRulesProps.showVerticalLines) !== null && _4 !== void 0 ? _4 : AxesAndRulesDefaults.showVerticalLines;
    var verticalLinesThickness = (_5 = axesAndRulesProps.verticalLinesThickness) !== null && _5 !== void 0 ? _5 : AxesAndRulesDefaults.verticalLinesThickness;
    var verticalLinesHeight = axesAndRulesProps.verticalLinesHeight;
    var verticalLinesColor = (_6 = axesAndRulesProps.verticalLinesColor) !== null && _6 !== void 0 ? _6 : AxesAndRulesDefaults.verticalLinesColor;
    var verticalLinesStrokeDashArray = (_7 = axesAndRulesProps.verticalLinesStrokeDashArray) !== null && _7 !== void 0 ? _7 : AxesAndRulesDefaults.verticalLinesStrokeDashArray;
    var verticalLinesShift = (_8 = axesAndRulesProps.verticalLinesShift) !== null && _8 !== void 0 ? _8 : AxesAndRulesDefaults.verticalLinesShift;
    var verticalLinesZIndex = (_9 = axesAndRulesProps.verticalLinesZIndex) !== null && _9 !== void 0 ? _9 : AxesAndRulesDefaults.verticalLinesZIndex;
    var verticalLinesSpacing = (_10 = axesAndRulesProps.verticalLinesSpacing) !== null && _10 !== void 0 ? _10 : AxesAndRulesDefaults.verticalLinesSpacing;
    var verticalLinesUptoDataPoint = (_11 = axesAndRulesProps.verticalLinesUptoDataPoint) !== null && _11 !== void 0 ? _11 : AxesAndRulesDefaults.verticalLinesUptoDataPoint;
    var verticalLinesStrokeLinecap = (_12 = axesAndRulesProps.verticalLinesStrokeLinecap) !== null && _12 !== void 0 ? _12 : AxesAndRulesDefaults.verticalLinesStrokeLinecap;
    var noOfVerticalLines = axesAndRulesProps.noOfVerticalLines;
    var secondaryXAxis = axesAndRulesProps.secondaryXAxis;
    var verticalLinesAr = noOfVerticalLines
        ? __spreadArray([], __read(Array(noOfVerticalLines).keys()), false) : (chartType === chartTypes.BUBBLE
        ? (xAxisLabelTexts !== null && xAxisLabelTexts !== void 0 ? xAxisLabelTexts : [])
        : (stackData !== null && stackData !== void 0 ? stackData : data)).map(function (item) {
        var showVerticalLine = item.showVerticalLine, verticalLineThickness = item.verticalLineThickness, verticalLineHeight = item.verticalLineHeight, verticalLineColor = item.verticalLineColor, verticalLineStrokeDashArray = item.verticalLineStrokeDashArray, verticalLineShift = item.verticalLineShift, verticalLineZIndex = item.verticalLineZIndex, verticalLineSpacing = item.verticalLineSpacing, verticalLineStrokeLinecap = item.verticalLineStrokeLinecap;
        return {
            showVerticalLine: showVerticalLine,
            verticalLineThickness: verticalLineThickness,
            verticalLineHeight: verticalLineHeight,
            verticalLineColor: verticalLineColor,
            verticalLineStrokeDashArray: verticalLineStrokeDashArray,
            verticalLineShift: verticalLineShift,
            verticalLineZIndex: verticalLineZIndex,
            verticalLineSpacing: verticalLineSpacing,
            verticalLineStrokeLinecap: verticalLineStrokeLinecap
        };
    });
    var extendedContainerHeight = containerHeight + overflowTop + 10;
    var dataSetArray = __spreadArray([], __read((dataSet !== null && dataSet !== void 0 ? dataSet : [])), false);
    var setWithMinValueInDataset = (_14 = (_13 = dataSetArray
        .filter(function (set) { return set.isSecondary; })).sort) === null || _14 === void 0 ? void 0 : _14.call(_13, function (a, b) {
        var minA = Math.min.apply(Math, __spreadArray([], __read(a.data.map(function (item) { return item.value; })), false));
        var minB = Math.min.apply(Math, __spreadArray([], __read(b.data.map(function (item) { return item.value; })), false));
        return minA - minB;
    })[0];
    var secondaryDataArrayWithMinValue = (secondaryData === null || secondaryData === void 0 ? void 0 : secondaryData.length)
        ? secondaryData
        : setWithMinValueInDataset === null || setWithMinValueInDataset === void 0 ? void 0 : setWithMinValueInDataset.data;
    var _23 = computeMaxAndMinItems(secondaryDataArrayWithMinValue, true, (_15 = secondaryYAxis === null || secondaryYAxis === void 0 ? void 0 : secondaryYAxis.roundToDigits) !== null && _15 !== void 0 ? _15 : roundToDigits, (_16 = secondaryYAxis === null || secondaryYAxis === void 0 ? void 0 : secondaryYAxis.showFractionalValues) !== null && _16 !== void 0 ? _16 : showFractionalValues), secondaryMaxItem = _23.maxItem, secondaryMinItem = _23.minItem;
    var containerHeightIncludingBelowXAxis = extendedContainerHeight + biggerNegativeYAxisHeight;
    var horizSectionProps = {
        chartType: chartType,
        width: width,
        horizSections: horizSections,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        totalWidth: totalWidth,
        endSpacing: endSpacing,
        yAxisSide: yAxisSide,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        sectionColors: sectionColors,
        stepHeight: stepHeight,
        negativeStepHeight: negativeStepHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        yAxisLabelContainerStyle: yAxisLabelContainerStyle,
        yAxisThickness: yAxisThickness,
        yAxisColor: yAxisColor,
        yAxisExtraHeight: yAxisExtraHeight,
        trimYAxisAtTop: trimYAxisAtTop,
        xAxisThickness: xAxisThickness,
        xAxisColor: xAxisColor,
        xAxisLength: xAxisLength,
        xAxisType: xAxisType,
        dashWidth: dashWidth,
        dashGap: dashGap,
        backgroundColor: backgroundColor,
        hideRules: hideRules,
        rulesLength: rulesLength,
        rulesType: rulesType,
        rulesThickness: rulesThickness,
        rulesColor: rulesColor,
        rulesConfigArray: rulesConfigArray,
        spacing: spacing,
        showYAxisIndices: showYAxisIndices,
        yAxisIndicesHeight: yAxisIndicesHeight,
        yAxisIndicesWidth: yAxisIndicesWidth,
        yAxisIndicesColor: yAxisIndicesColor,
        hideOrigin: hideOrigin,
        hideYAxisText: hideYAxisText,
        showFractionalValues: showFractionalValues,
        yAxisTextNumberOfLines: yAxisTextNumberOfLines,
        yAxisLabelPrefix: yAxisLabelPrefix,
        yAxisLabelSuffix: yAxisLabelSuffix,
        yAxisTextStyle: yAxisTextStyle,
        rotateYAxisTexts: rotateYAxisTexts,
        rtl: rtl,
        containerHeight: containerHeight,
        overflowTop: overflowTop,
        maxValue: maxValue,
        referenceLinesConfig: referenceLinesConfig,
        yAxisLabelTexts: yAxisLabelTexts,
        yAxisOffset: yAxisOffset,
        horizontal: horizontal,
        yAxisAtTop: yAxisAtTop,
        stepValue: stepValue,
        negativeStepValue: negativeStepValue,
        roundToDigits: roundToDigits,
        secondaryYAxis: secondaryYAxis,
        formatYLabel: axesAndRulesProps.formatYLabel,
        secondaryXAxis: secondaryXAxis,
        secondaryMaxItem: secondaryMaxItem,
        secondaryMinItem: secondaryMinItem,
        secondaryStepValue: secondaryStepValue,
        secondaryNegativeStepValue: secondaryNegativeStepValue,
        secondaryNoOfSectionsBelowXAxis: secondaryNoOfSectionsBelowXAxis,
        showSecondaryFractionalValues: showSecondaryFractionalValues,
        secondaryRoundToDigits: secondaryRoundToDigits,
        secondaryStepHeight: secondaryStepHeight,
        secondaryNegativeStepHeight: secondaryNegativeStepHeight,
        customBackground: props.customBackground,
        onlyPositive: onlyPositive,
        floatingYAxisLabels: floatingYAxisLabels,
        allowFontScaling: allowFontScaling
    };
    var lineInBarChartProps = {
        yAxisLabelWidth: yAxisLabelWidth,
        initialSpacing: initialSpacing,
        spacing: spacing,
        containerHeight: containerHeight,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        lineConfig: lineConfig,
        maxValue: (lineConfig === null || lineConfig === void 0 ? void 0 : lineConfig.isSecondary)
            ? ((_17 = secondaryYAxis === null || secondaryYAxis === void 0 ? void 0 : secondaryYAxis.maxValue) !== null && _17 !== void 0 ? _17 : maxValue)
            : maxValue,
        animatedWidth: animatedWidth,
        lineBehindBars: lineBehindBars,
        points: points,
        arrowPoints: arrowPoints,
        data: (lineData === null || lineData === void 0 ? void 0 : lineData.length) ? lineData : (stackData !== null && stackData !== void 0 ? stackData : data),
        totalWidth: totalWidth,
        barWidth: barWidth,
        labelsExtraHeight: labelsExtraHeight,
        scrollEventThrottle: scrollEventThrottle,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        selectedIndex: selectedIndex,
        yAxisOffset: (lineConfig === null || lineConfig === void 0 ? void 0 : lineConfig.isSecondary)
            ? ((_18 = secondaryYAxis === null || secondaryYAxis === void 0 ? void 0 : secondaryYAxis.yAxisOffset) !== null && _18 !== void 0 ? _18 : 0)
            : (yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0),
        strokeDashArray: (_19 = lineConfig === null || lineConfig === void 0 ? void 0 : lineConfig.strokeDashArray) !== null && _19 !== void 0 ? _19 : [0, 0],
        highlightEnabled: highlightEnabled,
        lowlightOpacity: lowlightOpacity,
        disableForeignObject: disableForeignObject
    };
    var lineInBarChartProps2 = __assign(__assign({}, lineInBarChartProps), { lineConfig: lineConfig2, points: points2, data: lineData2 !== null && lineData2 !== void 0 ? lineData2 : [], yAxisOffset: (lineConfig2 === null || lineConfig2 === void 0 ? void 0 : lineConfig2.isSecondary)
            ? ((_20 = secondaryYAxis === null || secondaryYAxis === void 0 ? void 0 : secondaryYAxis.yAxisOffset) !== null && _20 !== void 0 ? _20 : 0)
            : (yAxisOffset !== null && yAxisOffset !== void 0 ? yAxisOffset : 0) });
    var verticalLinesProps = {
        showVerticalLines: props.showVerticalLines,
        verticalLinesAr: verticalLinesAr,
        verticalLinesSpacing: verticalLinesSpacing,
        spacing: (_21 = lineConfig === null || lineConfig === void 0 ? void 0 : lineConfig.spacing) !== null && _21 !== void 0 ? _21 : spacing,
        initialSpacing: initialSpacing,
        verticalLinesZIndex: verticalLinesZIndex,
        verticalLinesHeight: verticalLinesHeight,
        verticalLinesThickness: verticalLinesThickness,
        verticalLinesColor: verticalLinesColor,
        verticalLinesStrokeDashArray: verticalLinesStrokeDashArray,
        verticalLinesShift: verticalLinesShift,
        verticalLinesUptoDataPoint: verticalLinesUptoDataPoint,
        verticalLinesStrokeLinecap: verticalLinesStrokeLinecap,
        xAxisThickness: xAxisThickness,
        labelsExtraHeight: labelsExtraHeight,
        containerHeight: containerHeight,
        data: data,
        stackData: stackData,
        barWidth: barWidth,
        maxValue: maxValue,
        chartType: chartType,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        yAxisLabelWidth: yAxisLabelWidth,
        totalWidth: totalWidth,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift
    };
    var actualContainerHeight = containerHeightIncludingBelowXAxis + labelsExtraHeight - 10;
    var actualContainerWidth = (width !== null && width !== void 0 ? width : totalWidth) + yAxisLabelWidth;
    /*******************************************************************************************************************************************/
    /** *************                                 horizontal chart related calculations                                   *******************/
    /*******************************************************************************************************************************************/
    var containerHeightIncludingXaxisLabels = actualContainerHeight + BarDefaults.labelsWidthForHorizontal;
    var difBwWidthHeight = actualContainerWidth - containerHeightIncludingXaxisLabels;
    var transformForHorizontal = [
        { rotate: rtl ? '-90deg' : '90deg' },
        {
            translateY: -shiftX + (rtl ? -difBwWidthHeight + 14 : difBwWidthHeight) / 2 - 20
        },
        {
            translateX: shiftY +
                (rtl
                    ? (props.width ? -98 - endSpacing : -75 - endSpacing) -
                        difBwWidthHeight
                    : props.width
                        ? difBwWidthHeight
                        : difBwWidthHeight - 40) /
                    2 +
                (yAxisAtTop ? (rtl ? (props.width ? 12 : 40) : 12) : 52)
        }
    ];
    var transformForHorizontalForReactJS = "rotate(".concat(rtl ? '-90deg' : '90deg', ")\n  translateY(").concat(-shiftX + (rtl ? -difBwWidthHeight + 14 : difBwWidthHeight) / 2 - 20, ")\n  translateX(").concat(shiftY +
        (rtl
            ? (props.width ? -98 - endSpacing : -75 - endSpacing) - difBwWidthHeight
            : props.width
                ? difBwWidthHeight
                : difBwWidthHeight - 40) /
            2 +
        (yAxisAtTop ? (rtl ? (props.width ? 12 : 40) : 12) : 52), ")");
    var _24 = __read(useState(false), 2), canMomentum = _24[0], setCanMomentum = _24[1];
    var isCloseToEnd = function (_a) {
        var layoutMeasurement = _a.layoutMeasurement, contentOffset = _a.contentOffset, contentSize = _a.contentSize;
        return isRTL
            ? contentOffset.x <= initialSpacing
            : layoutMeasurement.width + contentOffset.x >=
                contentSize.width - initialSpacing - endReachedOffset;
    };
    // const isCloseToStart = ({ layoutMeasurement, contentOffset }) => {
    //   return layoutMeasurement.width + contentOffset.x <= initialSpacing;
    // };
    var isCloseToStart = function (_a) {
        var layoutMeasurement = _a.layoutMeasurement, contentOffset = _a.contentOffset, contentSize = _a.contentSize;
        return isRTL
            ? layoutMeasurement.width + contentOffset.x >=
                contentSize.width - initialSpacing - endReachedOffset
            : contentOffset.x <= initialSpacing;
    };
    useEffect(function () {
        if (pointerConfig && getPointerProps) {
            getPointerProps({ pointerIndex: pointerIndex, pointerX: pointerX, pointerY: pointerY });
        }
    }, [pointerIndex, pointerX, pointerY]);
    return {
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        trimYAxisAtTop: trimYAxisAtTop,
        yAxisExtraHeight: yAxisExtraHeight,
        overflowTop: overflowTop,
        xAxisLabelsHeight: xAxisLabelsHeight,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        actualContainerWidth: actualContainerWidth,
        transformForHorizontal: transformForHorizontal,
        transformForHorizontalForReactJS: transformForHorizontalForReactJS,
        horizSectionProps: horizSectionProps,
        referenceLinesOverChartContent: referenceLinesOverChartContent,
        setCanMomentum: setCanMomentum,
        isCloseToStart: isCloseToStart,
        isCloseToEnd: isCloseToEnd,
        canMomentum: canMomentum,
        yAxisAtTop: yAxisAtTop,
        yAxisThickness: yAxisThickness,
        yAxisSide: yAxisSide,
        showVerticalLines: showVerticalLines,
        verticalLinesProps: verticalLinesProps,
        lineInBarChartProps: lineInBarChartProps,
        lineInBarChartProps2: lineInBarChartProps2
    };
};
