// gifted-charts-core/dist/LineChart/LineChartBiColor.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useEffect, useMemo, useState } from 'react';
import { AxesAndRulesDefaults, LineDefaults, chartTypes } from '../utils/constants';
import { getAxesAndRulesProps, getExtendedContainerHeightWithPadding } from '../utils';
var initialData = null;
export var useLineChartBiColor = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36, _37, _38, _39, _40, _41, _42, _43, _44, _45, _46, _47, _48, _49, _50, _51, _52, _53, _54;
    var _55 = __read(useState(false), 2), toggle = _55[0], setToggle = _55[1];
    var _56 = __read(useState([]), 2), pointsArray = _56[0], setPointsArray = _56[1];
    var _57 = __read(useState([]), 2), fillPointsArray = _57[0], setFillPointsArray = _57[1];
    var _58 = __read(useState(-1), 2), selectedIndex = _58[0], setSelectedIndex = _58[1];
    var containerHeight = (_a = props.height) !== null && _a !== void 0 ? _a : AxesAndRulesDefaults.containerHeight;
    var noOfSections = (_b = props.noOfSections) !== null && _b !== void 0 ? _b : AxesAndRulesDefaults.noOfSections;
    var data = useMemo(function () {
        if (!props.data) {
            return [];
        }
        if (props.yAxisOffset) {
            return props.data.map(function (item) {
                var _a;
                item.value = item.value - ((_a = props.yAxisOffset) !== null && _a !== void 0 ? _a : 0);
                return item;
            });
        }
        return props.data;
    }, [props.yAxisOffset, props.data]);
    var allowFontScaling = (_c = props.allowFontScaling) !== null && _c !== void 0 ? _c : AxesAndRulesDefaults.allowFontScaling;
    var disableForeignObject = (_d = props.disableForeignObject) !== null && _d !== void 0 ? _d : LineDefaults.disableForeignObject;
    var scrollToEnd = (_e = props.scrollToEnd) !== null && _e !== void 0 ? _e : LineDefaults.scrollToEnd;
    var scrollAnimation = (_f = props.scrollAnimation) !== null && _f !== void 0 ? _f : LineDefaults.scrollAnimation;
    var scrollEventThrottle = (_g = props.scrollEventThrottle) !== null && _g !== void 0 ? _g : LineDefaults.scrollEventThrottle;
    var labelsExtraHeight = (_h = props.labelsExtraHeight) !== null && _h !== void 0 ? _h : 0;
    var animationDuration = (_j = props.animationDuration) !== null && _j !== void 0 ? _j : LineDefaults.animationDuration;
    var startIndex1 = (_k = props.startIndex) !== null && _k !== void 0 ? _k : 0;
    var endIndex1;
    if (props.endIndex === undefined || props.endIndex === null) {
        endIndex1 = data.length - 1;
    }
    else {
        endIndex1 = props.endIndex;
    }
    if (!initialData) {
        initialData = __spreadArray([], __read(data), false);
    }
    var adjustToWidth = (_l = props.adjustToWidth) !== null && _l !== void 0 ? _l : false;
    var initialSpacing = (_m = props.initialSpacing) !== null && _m !== void 0 ? _m : LineDefaults.initialSpacing;
    var endSpacing = (_o = props.endSpacing) !== null && _o !== void 0 ? _o : (adjustToWidth ? 0 : LineDefaults.endSpacing);
    var thickness = (_p = props.thickness) !== null && _p !== void 0 ? _p : LineDefaults.thickness;
    var spacing = (_q = props.spacing) !== null && _q !== void 0 ? _q : (adjustToWidth
        ? (((_r = props.width) !== null && _r !== void 0 ? _r : AxesAndRulesDefaults.width) - initialSpacing) /
            data.length
        : LineDefaults.spacing);
    var xAxisThickness = (_s = props.xAxisThickness) !== null && _s !== void 0 ? _s : AxesAndRulesDefaults.xAxisThickness;
    var dataPointsHeight1 = (_t = props.dataPointsHeight) !== null && _t !== void 0 ? _t : LineDefaults.dataPointsHeight;
    var dataPointsWidth1 = (_u = props.dataPointsWidth) !== null && _u !== void 0 ? _u : LineDefaults.dataPointsWidth;
    var dataPointsRadius1 = (_v = props.dataPointsRadius) !== null && _v !== void 0 ? _v : LineDefaults.dataPointsRadius;
    var dataPointsColor1 = (_w = props.dataPointsColor) !== null && _w !== void 0 ? _w : LineDefaults.dataPointsColor;
    var dataPointsShape1 = (_x = props.dataPointsShape) !== null && _x !== void 0 ? _x : LineDefaults.dataPointsShape;
    var areaChart = (_y = props.areaChart) !== null && _y !== void 0 ? _y : false;
    var textFontSize1 = (_z = props.textFontSize) !== null && _z !== void 0 ? _z : LineDefaults.textFontSize;
    var textColor1 = (_0 = props.textColor) !== null && _0 !== void 0 ? _0 : LineDefaults.textColor;
    var totalWidth = initialSpacing + endSpacing;
    var maxItem = 0;
    var minItem = 0;
    data.forEach(function (item, index) {
        if (item.value > maxItem) {
            maxItem = item.value;
        }
        if (item.value < minItem) {
            minItem = item.value;
        }
        totalWidth += index === data.length - 1 ? 0 : spacing;
    });
    if ((_1 = props.showFractionalValues) !== null && _1 !== void 0 ? _1 : props.roundToDigits) {
        maxItem *= 10 * ((_2 = props.roundToDigits) !== null && _2 !== void 0 ? _2 : 1);
        maxItem = maxItem + (10 - (maxItem % 10));
        maxItem /= 10 * ((_3 = props.roundToDigits) !== null && _3 !== void 0 ? _3 : 1);
        maxItem = parseFloat(maxItem.toFixed((_4 = props.roundToDigits) !== null && _4 !== void 0 ? _4 : 1));
        if (minItem !== 0) {
            minItem *= 10 * ((_5 = props.roundToDigits) !== null && _5 !== void 0 ? _5 : 1);
            minItem = minItem - (10 + (minItem % 10));
            minItem /= 10 * ((_6 = props.roundToDigits) !== null && _6 !== void 0 ? _6 : 1);
            minItem = parseFloat(minItem.toFixed((_7 = props.roundToDigits) !== null && _7 !== void 0 ? _7 : 1));
        }
    }
    else {
        maxItem = maxItem + (10 - (maxItem % 10));
        if (minItem !== 0) {
            minItem = minItem - (10 + (minItem % 10));
        }
    }
    var maxValue = (_8 = props.maxValue) !== null && _8 !== void 0 ? _8 : maxItem;
    var mostNegativeValue = (_9 = props.mostNegativeValue) !== null && _9 !== void 0 ? _9 : minItem;
    var extendedContainerHeight = getExtendedContainerHeightWithPadding(containerHeight, props.overflowTop);
    var yAtxAxis = extendedContainerHeight - xAxisThickness / 2;
    var getX = function (index) { return initialSpacing + spacing * index; };
    var getY = function (index) {
        return yAtxAxis - (data[index].value * containerHeight) / maxValue;
    };
    useEffect(function () {
        var ppArray = [];
        var pp = 'M' + initialSpacing + ' ' + getY(0);
        var prevValuev;
        var nextValue;
        for (var i_1 = 0; i_1 < data.length - 1; i_1++) {
            prevValuev = data[i_1].value;
            nextValue = data[i_1 + 1].value;
            if (prevValuev < 0 && nextValue < 0) {
                pp += 'L' + getX(i_1) + ' ' + getY(i_1) + ' ';
            }
            else if (prevValuev < 0 && nextValue > 0) {
                pp += 'L' + getX(i_1) + ' ' + getY(i_1) + ' ';
                var prevX = getX(i_1);
                var prevY = getY(i_1);
                var nextX = getX(i_1 + 1);
                var nextY = getY(i_1 + 1);
                var slope = (nextY - prevY) / (nextX - prevX);
                var x = (yAtxAxis - prevY) / slope + prevX;
                pp += 'L' + (x - thickness / 2) + ' ' + yAtxAxis + ' ';
                var pointsOb_1 = {
                    points: pp.startsWith('L') ? pp.replace('L', 'M') : pp,
                    color: 'red'
                };
                ppArray.push(pointsOb_1);
                setPointsArray(__spreadArray([], __read(ppArray), false));
                pp = 'M' + x + ' ' + yAtxAxis + ' L' + nextX + ' ' + nextY + ' ';
                pointsOb_1 = {
                    points: pp,
                    color: 'green'
                };
                ppArray.push(pointsOb_1);
            }
            else if (prevValuev > 0 && nextValue < 0) {
                pp += 'L' + getX(i_1) + ' ' + getY(i_1) + ' ';
                var prevX = getX(i_1);
                var prevY = getY(i_1);
                var nextX = getX(i_1 + 1);
                var nextY = getY(i_1 + 1);
                var slope = (nextY - prevY) / (nextX - prevX);
                var x = (yAtxAxis - prevY) / slope + prevX;
                pp += 'L' + (x - thickness / 2) + ' ' + yAtxAxis + ' ';
                var pointsOb_2 = {
                    points: pp.startsWith('L') ? pp.replace('L', 'M') : pp,
                    color: 'green'
                };
                ppArray.push(pointsOb_2);
                setPointsArray(__spreadArray([], __read(ppArray), false));
                pp = 'M' + x + ' ' + yAtxAxis + ' L' + nextX + ' ' + nextY + ' ';
                pointsOb_2 = {
                    points: pp,
                    color: 'red'
                };
                ppArray.push(pointsOb_2);
            }
            else {
                pp += 'L' + getX(i_1) + ' ' + getY(i_1) + ' ';
            }
        }
        var i = data.length - 1;
        prevValuev = data[i - 1].value;
        nextValue = data[i].value;
        if ((prevValuev > 0 && nextValue > 0) ||
            (prevValuev < 0 && nextValue < 0)) {
            pp += 'L' + getX(i) + ' ' + getY(i) + ' ';
        }
        var pointsOb = {
            points: pp.startsWith('L') ? pp.replace('L', 'M') : pp,
            color: nextValue > 0 ? 'green' : 'red'
        };
        ppArray.push(pointsOb);
        setPointsArray(__spreadArray([], __read(ppArray), false));
        /** *************************          For Area Charts          *************************/
        var startIndex = -1;
        var endIndex = -1;
        var startX;
        var startY;
        var endY;
        var color = 'green';
        var localArray = [];
        var broken = false;
        pp = 'M' + initialSpacing + ' ' + yAtxAxis;
        for (i = 0; i < data.length - 1; i++) {
            prevValuev = data[i].value;
            nextValue = data[i + 1].value;
            pp += 'L' + getX(i) + ' ' + getY(i) + ' ';
            if ((prevValuev > 0 && nextValue < 0) ||
                (prevValuev < 0 && nextValue > 0)) {
                var prevX = getX(i);
                var prevY = getY(i);
                var nextX = getX(i + 1);
                var nextY = getY(i + 1);
                var slope = (nextY - prevY) / (nextX - prevX);
                var x = (yAtxAxis - prevY) / slope + prevX;
                pp += 'L' + (x - thickness / 2) + ' ' + yAtxAxis + ' ';
                broken = true;
                break;
            }
        }
        if (!broken) {
            i = data.length - 1;
            pp +=
                'L' +
                    getX(i) +
                    ' ' +
                    getY(i) +
                    ' L' +
                    getX(i) +
                    ' ' +
                    (yAtxAxis - xAxisThickness / 2);
        }
        localArray.push({
            points: pp,
            color: data[0].value >= 0 ? 'green' : 'red'
        });
        var xs = [];
        data.forEach(function (item, index) {
            var x = getX(index);
            xs.push(x + '');
        });
        pointsArray.forEach(function (item, index) {
            var splitArray = item.points
                .split(' ')
                .filter(function (spItem) { return spItem && spItem !== ' '; });
            if (splitArray[1] === yAtxAxis + '' &&
                !xs.includes(splitArray[0].replace('M', '').replace('L', ''))) {
                startIndex = index;
                startX = splitArray[0].replace('M', '').replace('L', '');
                if (splitArray.length > 3) {
                    startY = splitArray[1].replace('M', '').replace('L', '');
                    endY = splitArray[3].replace('M', '').replace('L', '');
                    if (Number(startY) < Number(endY)) {
                        color = 'red';
                    }
                    else {
                        color = 'green';
                    }
                }
            }
            if (splitArray[splitArray.length - 1] === yAtxAxis + '' &&
                !xs.includes(splitArray[splitArray.length - 2].replace('M', '').replace('L', ''))) {
                endIndex = index;
            }
            if (startX) {
                var filPts = '';
                for (var j = startIndex; j <= endIndex; j++) {
                    if (pointsArray[j]) {
                        filPts += pointsArray[j].points.replaceAll('M', 'L');
                    }
                }
                filPts += 'L ' + startX + ' ' + yAtxAxis;
                localArray.push({ points: filPts.replace('L', 'M'), color: color });
            }
        });
        if (broken) {
            pp = 'M' + getX(data.length - 1) + ' ' + yAtxAxis;
            for (var i_2 = data.length - 1; i_2 > 0; i_2--) {
                prevValuev = data[i_2].value;
                nextValue = data[i_2 - 1].value;
                pp += 'L' + getX(i_2) + ' ' + getY(i_2) + ' ';
                if ((prevValuev > 0 && nextValue < 0) ||
                    (prevValuev < 0 && nextValue > 0)) {
                    var prevX = getX(i_2);
                    var prevY = getY(i_2);
                    var nextX = getX(i_2 - 1);
                    var nextY = getY(i_2 - 1);
                    var slope = (nextY - prevY) / (nextX - prevX);
                    var x = (yAtxAxis - prevY) / slope + prevX;
                    pp += 'L' + x + ' ' + yAtxAxis + ' ';
                    break;
                }
            }
            localArray.push({
                points: pp,
                color: data[data.length - 1].value > 0 ? 'green' : 'red'
            });
        }
        setFillPointsArray(__spreadArray([], __read(localArray), false));
        setToggle(true);
    }, [
        areaChart,
        containerHeight,
        data,
        dataPointsWidth1,
        initialSpacing,
        spacing,
        xAxisThickness,
        toggle,
        maxValue
    ]);
    var horizSections = [{ value: '0' }];
    var stepHeight = (_10 = props.stepHeight) !== null && _10 !== void 0 ? _10 : containerHeight / noOfSections;
    var stepValue = (_11 = props.stepValue) !== null && _11 !== void 0 ? _11 : maxValue / noOfSections;
    var noOfSectionsBelowXAxis = (_12 = props.noOfSectionsBelowXAxis) !== null && _12 !== void 0 ? _12 : Math.round(Math.ceil(-mostNegativeValue / stepValue));
    var thickness1 = (_13 = props.thickness) !== null && _13 !== void 0 ? _13 : LineDefaults.thickness;
    var zIndex = (_14 = props.zIndex) !== null && _14 !== void 0 ? _14 : 0;
    var strokeDashArray1 = props.strokeDashArray;
    var rotateLabel = (_15 = props.rotateLabel) !== null && _15 !== void 0 ? _15 : AxesAndRulesDefaults.rotateLabel;
    var isAnimated = (_16 = props.isAnimated) !== null && _16 !== void 0 ? _16 : LineDefaults.isAnimated;
    var hideDataPoints1 = (_17 = props.hideDataPoints) !== null && _17 !== void 0 ? _17 : LineDefaults.hideDataPoints;
    var color = (_18 = props.color) !== null && _18 !== void 0 ? _18 : 'green';
    var colorNegative = (_19 = props.colorNegative) !== null && _19 !== void 0 ? _19 : 'red';
    var startFillColor = (_20 = props.startFillColor) !== null && _20 !== void 0 ? _20 : 'lightgreen';
    var endFillColor = (_21 = props.endFillColor) !== null && _21 !== void 0 ? _21 : 'white';
    var startOpacity = (_22 = props.startOpacity) !== null && _22 !== void 0 ? _22 : LineDefaults.startOpacity;
    var endOpacity = (_23 = props.endOpacity) !== null && _23 !== void 0 ? _23 : LineDefaults.endOpacity;
    var startFillColorNegative = (_24 = props.startFillColorNegative) !== null && _24 !== void 0 ? _24 : 'pink';
    var endFillColorNegative = (_25 = props.endFillColorNegative) !== null && _25 !== void 0 ? _25 : 'white';
    var startOpacityNegative = (_26 = props.startOpacityNegative) !== null && _26 !== void 0 ? _26 : LineDefaults.startOpacity;
    var endOpacityNegative = (_27 = props.endOpacityNegative) !== null && _27 !== void 0 ? _27 : LineDefaults.endOpacity;
    var gradientDirection = (_28 = props.gradientDirection) !== null && _28 !== void 0 ? _28 : 'vertical';
    var showXAxisIndices = (_29 = props.showXAxisIndices) !== null && _29 !== void 0 ? _29 : AxesAndRulesDefaults.showXAxisIndices;
    var xAxisIndicesHeight = (_30 = props.xAxisIndicesHeight) !== null && _30 !== void 0 ? _30 : AxesAndRulesDefaults.xAxisIndicesHeight;
    var xAxisIndicesWidth = (_31 = props.xAxisIndicesWidth) !== null && _31 !== void 0 ? _31 : AxesAndRulesDefaults.xAxisIndicesWidth;
    var xAxisIndicesColor = (_32 = props.xAxisIndicesColor) !== null && _32 !== void 0 ? _32 : AxesAndRulesDefaults.xAxisIndicesColor;
    var xAxisTextNumberOfLines = (_33 = props.xAxisTextNumberOfLines) !== null && _33 !== void 0 ? _33 : AxesAndRulesDefaults.xAxisTextNumberOfLines;
    var horizontalRulesStyle = props.horizontalRulesStyle;
    var showFractionalValues = (_34 = props.showFractionalValues) !== null && _34 !== void 0 ? _34 : AxesAndRulesDefaults.showFractionalValues;
    var yAxisLabelWidth = (_35 = props.yAxisLabelWidth) !== null && _35 !== void 0 ? _35 : (props.hideYAxisText
        ? AxesAndRulesDefaults.yAxisEmptyLabelWidth
        : AxesAndRulesDefaults.yAxisLabelWidth);
    var yAxisExtraHeightAtTop = props.trimYAxisAtTop
        ? 0
        : ((_36 = props.yAxisExtraHeight) !== null && _36 !== void 0 ? _36 : containerHeight / 20);
    var horizontal = false;
    var yAxisAtTop = false;
    var disableScroll = (_37 = props.disableScroll) !== null && _37 !== void 0 ? _37 : LineDefaults.disableScroll;
    var showScrollIndicator = (_38 = props.showScrollIndicator) !== null && _38 !== void 0 ? _38 : LineDefaults.showScrollIndicator;
    var focusEnabled = (_39 = props.focusEnabled) !== null && _39 !== void 0 ? _39 : LineDefaults.focusEnabled;
    var showDataPointOnFocus = (_40 = props.showDataPointOnFocus) !== null && _40 !== void 0 ? _40 : LineDefaults.showDataPointOnFocus;
    var showStripOnFocus = (_41 = props.showStripOnFocus) !== null && _41 !== void 0 ? _41 : LineDefaults.showStripOnFocus;
    var showTextOnFocus = (_42 = props.showTextOnFocus) !== null && _42 !== void 0 ? _42 : LineDefaults.showTextOnFocus;
    var stripHeight = props.stripHeight;
    var stripWidth = (_43 = props.stripWidth) !== null && _43 !== void 0 ? _43 : LineDefaults.stripWidth;
    var stripColor = (_44 = props.stripColor) !== null && _44 !== void 0 ? _44 : color;
    var stripOpacity = (_45 = props.stripOpacity) !== null && _45 !== void 0 ? _45 : (startOpacity + endOpacity) / 2;
    var unFocusOnPressOut = (_46 = props.unFocusOnPressOut) !== null && _46 !== void 0 ? _46 : LineDefaults.unFocusOnPressOut;
    var delayBeforeUnFocus = (_47 = props.delayBeforeUnFocus) !== null && _47 !== void 0 ? _47 : LineDefaults.delayBeforeUnFocus;
    horizSections.pop();
    for (var i = 0; i <= noOfSections; i++) {
        var value = maxValue - stepValue * i;
        if ((_48 = props.showFractionalValues) !== null && _48 !== void 0 ? _48 : props.roundToDigits) {
            value = parseFloat(value.toFixed((_49 = props.roundToDigits) !== null && _49 !== void 0 ? _49 : 1));
        }
        horizSections.push({
            value: props.yAxisLabelTexts
                ? ((_50 = props.yAxisLabelTexts[noOfSections - i]) !== null && _50 !== void 0 ? _50 : value.toString())
                : value.toString()
        });
    }
    var dataPointsRadius = (_51 = props.dataPointsRadius) !== null && _51 !== void 0 ? _51 : LineDefaults.dataPointsRadius;
    var dataPointsWidth = (_52 = props.dataPointsWidth) !== null && _52 !== void 0 ? _52 : LineDefaults.dataPointsWidth;
    var extraWidthDueToDataPoint = props.hideDataPoints
        ? 0
        : (dataPointsRadius !== null && dataPointsRadius !== void 0 ? dataPointsRadius : dataPointsWidth);
    var barAndLineChartsWrapperProps = {
        chartType: chartTypes.LINE_BI_COLOR,
        containerHeight: containerHeight,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: stepHeight,
        negativeStepHeight: stepHeight,
        labelsExtraHeight: labelsExtraHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        rtl: false,
        shiftX: 0,
        shiftY: 0,
        yAxisAtTop: yAxisAtTop,
        initialSpacing: initialSpacing,
        data: data,
        stackData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        secondaryData: [],
        barWidth: 0, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        xAxisThickness: xAxisThickness,
        totalWidth: totalWidth,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollToIndex: props.scrollToIndex,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        indicatorColor: props.indicatorColor,
        selectedIndex: [selectedIndex],
        setSelectedIndex: setSelectedIndex,
        spacing: spacing,
        showLine: false,
        lineConfig: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineConfig2: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        maxValue: maxValue,
        lineData: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineData2: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineBehindBars: false,
        points: pointsArray,
        points2: '', // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        arrowPoints: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        remainingScrollViewProps: {},
        // horizSectionProps-
        width: props.width,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        showFractionalValues: showFractionalValues,
        axesAndRulesProps: getAxesAndRulesProps(props, containerHeight, stepHeight, stepValue, noOfSections, (_53 = props.roundToDigits) !== null && _53 !== void 0 ? _53 : 0, stepValue, 0, 0, false, 0),
        yAxisLabelTexts: props.yAxisLabelTexts,
        yAxisOffset: props.yAxisOffset,
        rotateYAxisTexts: 0,
        hideAxesAndRules: props.hideAxesAndRules,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        // These are Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        pointerConfig: undefined,
        getPointerProps: null,
        pointerIndex: 0,
        pointerX: 0,
        pointerY: 0,
        endReachedOffset: (_54 = props.endReachedOffset) !== null && _54 !== void 0 ? _54 : LineDefaults.endReachedOffset,
        extraWidthDueToDataPoint: extraWidthDueToDataPoint,
        highlightEnabled: LineDefaults.highlightEnabled,
        lowlightOpacity: LineDefaults.lowlightOpacity,
        xAxisLabelsAtBottom: false,
        onScrollEndDrag: props.onScrollEndDrag,
        floatingYAxisLabels: props.floatingYAxisLabels,
        allowFontScaling: allowFontScaling,
        showVerticalLines: props.showVerticalLines,
        disableForeignObject: disableForeignObject
    };
    return {
        toggle: toggle,
        setToggle: setToggle,
        pointsArray: pointsArray,
        setPointsArray: setPointsArray,
        fillPointsArray: fillPointsArray,
        setFillPointsArray: setFillPointsArray,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        containerHeight: containerHeight,
        noOfSections: noOfSections,
        data: data,
        scrollToEnd: scrollToEnd,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        labelsExtraHeight: labelsExtraHeight,
        animationDuration: animationDuration,
        startIndex1: startIndex1,
        endIndex1: endIndex1,
        initialData: initialData,
        adjustToWidth: adjustToWidth,
        initialSpacing: initialSpacing,
        endSpacing: endSpacing,
        thickness: thickness,
        spacing: spacing,
        xAxisThickness: xAxisThickness,
        dataPointsHeight1: dataPointsHeight1,
        dataPointsWidth1: dataPointsWidth1,
        dataPointsRadius1: dataPointsRadius1,
        dataPointsColor1: dataPointsColor1,
        dataPointsShape1: dataPointsShape1,
        areaChart: areaChart,
        textFontSize1: textFontSize1,
        textColor1: textColor1,
        totalWidth: totalWidth,
        maxItem: maxItem,
        minItem: minItem,
        maxValue: maxValue,
        mostNegativeValue: mostNegativeValue,
        extendedContainerHeight: extendedContainerHeight,
        getX: getX,
        getY: getY,
        yAtxAxis: yAtxAxis,
        stepHeight: stepHeight,
        stepValue: stepValue,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        thickness1: thickness1,
        zIndex: zIndex,
        strokeDashArray1: strokeDashArray1,
        rotateLabel: rotateLabel,
        isAnimated: isAnimated,
        hideDataPoints1: hideDataPoints1,
        color: color,
        colorNegative: colorNegative,
        startFillColor: startFillColor,
        endFillColor: endFillColor,
        startOpacity: startOpacity,
        endOpacity: endOpacity,
        startFillColorNegative: startFillColorNegative,
        endFillColorNegative: endFillColorNegative,
        startOpacityNegative: startOpacityNegative,
        endOpacityNegative: endOpacityNegative,
        gradientDirection: gradientDirection,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        horizontalRulesStyle: horizontalRulesStyle,
        showFractionalValues: showFractionalValues,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        yAxisAtTop: yAxisAtTop,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        focusEnabled: focusEnabled,
        showDataPointOnFocus: showDataPointOnFocus,
        showStripOnFocus: showStripOnFocus,
        showTextOnFocus: showTextOnFocus,
        stripHeight: stripHeight,
        stripWidth: stripWidth,
        stripColor: stripColor,
        stripOpacity: stripOpacity,
        unFocusOnPressOut: unFocusOnPressOut,
        delayBeforeUnFocus: delayBeforeUnFocus,
        horizSections: horizSections,
        barAndLineChartsWrapperProps: barAndLineChartsWrapperProps,
        yAxisExtraHeightAtTop: yAxisExtraHeightAtTop,
        allowFontScaling: allowFontScaling
    };
};
